---
layout: page
title: Installation
permalink: /installation/
---

Table of Contents
=================

  * [Prerequisites](#prerequisites)
  * [Data preparation](#data-preparation)

## Prerequisites

[ICoVeR]({{ site.baseurl }}/) is designed and implemented as a regular [R](http://www.r-project.org/) package.
This guide was tested with [R](http://www.r-project.org/) v3.2.0 and [RStudio](http://www.rstudio.com/) v0.98.1102.
Newer versions should work, please report any problems at [our github page](https://github.com/bbroeksema/ICoVeR/issues).
Further more, a modern browser is required.
We have had most convenient experience with Chrome, but recent versions of Firefox worked as well.

To run our interactive interface, the OpenCPU R-package must be installed.
To this end run the following command in R or RStudio.

{% highlight r %}
# Bioconductor is required for the preprocessing of the fasta files
> source("http://bioconductor.org/biocLite.R")
> biocLite("Biostrings")

# OpenCPU is required to run the interactive binning tool
> install.packages("opencpu")

# devtools is required to install the interactive binning tool
> install.packages(devtools)

{% endhighlight %}

To get started with the guide you will have to download our [prepared package](link.to.package) or clone the [github repository](github.com).
In the remainder of this guide we will assume that the package is extracted in
your home directory: `~/icobire-guide/`.

## Data preparation

In order to get you started quickly with the tool, we [provide]() to prepared datasets.
The first data set is the one generated by [Wrighton et al](http://www.sciencemag.org/content/337/6102/1661).
The second data set, is the CSTR data set, which is generated by us from an anaerobic digester.

## Preprocessing

In order to load the data into our interactive contig binning system, we need to
pre-process the fasta and abundance level files. This is done by using the
scripts provided in the R.preprocessing directory. In this example we will
prepare the cstr data set for interactive contig binning.

The `PrepareDataForInteractiveBinning` function performs the following steps:

1. Extract gc_content and contig length for each contig from the fasta file
1. Extract tetra nucleotide frequencies for each contig from the fasta file
1. Combine extracted information with sample abundance levels into the files
   which are expected by the interactive contig binning tool.

**NOTE:** The ids in the fasta file must match the names in the abundance file.
Thus, if a contig in the fasta file starts with a line `>contig_123`, there must
be a line in the abundance file as where the name column has value `contig_123`.
Some error checking is done, but you should not rely on this and make sure that
the fasta file and the abundance file are prepared properly.

Start R studio and type the following commands.

{% highlight r %}
# Set the working directory (change path to your local checkout location)
> setwd("~/ICoVeR")

# Source the required R file for preprocessing. This will load all the required
# dependencies as well.
> source("R.preprocessing/preprocessing.R")
> PrepareDataForInteractiveBinning(
    dataset.name = "cstr",
    file.fasta  = "data//cstr_assembled.fasta",
    file.abundance = "data//cstr_avg_coverage.csv",
    dir.result = "data//prepared"
  )

## Installation

# Check if you find cstr.rda and cstr.schema.rda in: R.ICoVeR/data. If so, we
# continue installing the interactive binning application.
> library(devtools)

# NOTE: Before installing you should check the file R.ICoBiRe/R/sqlite.R. At the
#       top there is a variable declared named: p.db.dataset. The value for this
#       variable must match the data set you want to analyze (i.e. "cstr" in
#       this case, which is the default).
#
#       If you want to bin a different contig set, you must change this value
#       **before** installing the R.ICoBiRe package.
> install_local(file.path(getwd(), "R.ICoVeR"))
{% endhighlight %}

## Interactive binning

At this point everything is set up to start the interactive binning process.
In order to start the application we need to start OpenCPU:

{% highlight r %}
> library(opencpu)
Initiating OpenCPU server...
Using config: ~/.opencpu.conf
OpenCPU started.
[httpuv] http://localhost:2110/ocpu
OpenCPU single-user server ready.
{% endhighlight %}

If OpenCPU started without errors, the interactive application can now be accessed
in your browser at the following url:

    http://localhost:2110/ocpu/library/ICoVeR/www/

**NOTE 1:** The port number (i.e. 2110) must match with the output of OpenCPU in your RStudio session.
This can differ everytime you start the application.

**NOTE 2:** The application stores the most important data resulting from your analysis (i.e. clustering results and tags).
An analysis can therefore be splitted into several sessions.
The initial preparations are not required after the first (assuming binning of the same data set is continued).
To continue a previously stopped session, just start R(Studio), load the OpenCPU library as shown above, and point your browser to the correct url.
