<!DOCTYPE html>
<html>

  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width initial-scale=1" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>Overview</title>
    <meta name="description" content="This site provides data, scripts and a tutorial to replicate the contig binning results presented in [](PAPER titel).
">

    <link rel="stylesheet" href="/css/main.css">
    <link rel="canonical" href="http://bbroeksema.github.io/">
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/">Interactive Contig Binning</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/about/">About</a>
          
        
          
        
          
          <a class="page-link" href="/index.html">Overview</a>
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <h2 id="introduction">Introduction</h2>

<h2 id="required-software">Required Software</h2>

<p>This guide was tested with <a href="http://www.r-project.org/">R</a> v3.2.0 and
<a href="http://www.rstudio.com/">RStudio</a> v0.98.1102. Newer versions should work,
please report any problems at <a href="http://no.link.yet">TODO</a>. Further more, a modern
browser is required. We have had most convenient experience with Chrome, but
recent versions of Firefox worked as well.</p>

<p>To run our interactive interface, the OpenCPU R-package must be installed. To
this end run the following command in R or RStudio.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Bioconductor is required for the preprocessing of the fasta files</span>
<span class="o">&gt;</span> <span class="kn">source</span><span class="p">(</span><span class="s">&quot;http://bioconductor.org/biocLite.R&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> biocLite<span class="p">(</span><span class="s">&quot;Biostrings&quot;</span><span class="p">)</span>

<span class="c1"># OpenCPU is required to run the interactive binning tool</span>
<span class="o">&gt;</span> install.packages<span class="p">(</span><span class="s">&quot;opencpu&quot;</span><span class="p">)</span>

<span class="c1"># devtools is required to install the interactive binning tool</span>
<span class="o">&gt;</span> install.packages<span class="p">(</span>devtools<span class="p">)</span></code></pre></div>

<p>To get started with the guide you will have to download our
<a href="link.to.package">prepared package</a> or clone the <a href="github.com">github repository</a>.
In the remainder of this guide we will assume that the package is extracted in
your home directory: <code>~/icobire-guide/</code>.</p>

<h2 id="data">Data</h2>

<p>We provide the two data sets that have been analyzed using our interactive contig
binning system. The first data set is the one generated by
<a href="http://www.sciencemag.org/content/337/6102/1661">Wrighton et al</a>. The second
data set, is the CSTR data set, which is generated by us from an anaerobic
digester.</p>

<h2 id="assembly">Assembly</h2>

<p><strong>TODO</strong>: Currently the data contains a fasta file. I assume this file is
generated somehow from a fastq(?) file.</p>

<h2 id="abundance-extraction">Abundance extraction</h2>

<p><strong>TODO</strong>: Currently the data contains a csv fil with abundance levels. Weâ€™ll
need to describe how we generated them.</p>

<h2 id="preprocessing">Preprocessing</h2>

<p>In order to load the data into our interactive contig binning system, we need to
pre-process the fasta and abundance level files. This is done by using the
scripts provided in the R.preprocessing directory. In this example we will
prepare the cstr data set for interactive contig binning.</p>

<p>The <code>PrepareDataForInteractiveBinning</code> function performs the following steps:</p>

<ol>
  <li>Extract gc_content and contig length for each contig from the fasta file</li>
  <li>Extract tetra nucleotide frequencies for each contig from the fasta file</li>
  <li>Combine extracted information with sample abundance levels into the files
which are expected by the interactive contig binning tool.</li>
</ol>

<p><strong>NOTE:</strong> The ids in the fasta file must match the names in the abundance file.
Thus, if a contig in the fasta file starts with a line <code>&gt;contig_123</code>, there must
be a line in the abundance file as where the name column has value <code>contig_123</code>.
Some error checking is done, but you should not rely on this and make sure that
the fasta file and the abundance file are prepared properly.</p>

<p>Start R studio and type the following commands.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Set the working directory</span>
<span class="o">&gt;</span> <span class="kp">setwd</span><span class="p">(</span><span class="s">&quot;~/icobire-guide&quot;</span><span class="p">)</span>

<span class="c1"># Source the required R file for preprocessing. This will load all the required</span>
<span class="c1"># dependencies as well.</span>
<span class="o">&gt;</span> <span class="kn">source</span><span class="p">(</span><span class="s">&quot;R.preprocessing/preprocessing.R&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> PrepareDataForInteractiveBinning<span class="p">(</span>
    dataset.name <span class="o">=</span> <span class="s">&quot;cstr&quot;</span><span class="p">,</span>
    file.fasta  <span class="o">=</span> <span class="s">&quot;data//cstr_assembled.fasta&quot;</span><span class="p">,</span>
    file.abundance <span class="o">=</span> <span class="s">&quot;data//cstr_avg_coverage.csv&quot;</span><span class="p">,</span>
    dir.result <span class="o">=</span> <span class="s">&quot;data//prepared&quot;</span>
  <span class="p">)</span>

<span class="c1"># Check if you find cstr.rda and cstr.schema.rda in: R.ICoBiRe/data. If so, we</span>
<span class="c1"># continue installing the interactive binning application.</span>
<span class="o">&gt;</span> <span class="kn">library</span><span class="p">(</span>devtools<span class="p">)</span>

<span class="c1"># NOTE: Before installing you should check the file R.ICoBiRe/R/sqlite.R. At the</span>
<span class="c1">#       top there is a variable declared named: p.db.dataset. The value for this</span>
<span class="c1">#       variable must match the data set you want to analyze (i.e. &quot;cstr&quot; in</span>
<span class="c1">#       this case, which is the default).</span>
<span class="c1">#</span>
<span class="c1">#       If you want to bin a different contig set, you must change this value</span>
<span class="c1">#       **before** installing the R.ICoBiRe package.</span>
<span class="o">&gt;</span> install_local<span class="p">(</span><span class="kp">file.path</span><span class="p">(</span><span class="kp">getwd</span><span class="p">(),</span> <span class="s">&quot;R.ICoBiRe&quot;</span><span class="p">))</span></code></pre></div>

<h2 id="interactive-binning">Interactive binning</h2>

<p>At this point everything is set up to start the interactive binning process.
In order to start the application we need to start OpenCPU:</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="o">&gt;</span> <span class="kn">library</span><span class="p">(</span>opencpu<span class="p">)</span>
Initiating OpenCPU server...
Using config<span class="o">:</span> <span class="o">~/</span><span class="m">.</span>opencpu.conf
OpenCPU started.
<span class="p">[</span>httpuv<span class="p">]</span> http<span class="o">://</span>localhost<span class="o">:</span><span class="m">2110</span><span class="o">/</span>ocpu
OpenCPU <span class="kp">single</span><span class="o">-</span>user server ready.</code></pre></div>

<p>If OpenCPU started without errors, the interactive application can now be accessed
in your browser at the following url:</p>

<pre><code>http://localhost:2110/ocpu/library/ICoBiRe/www/
</code></pre>

<p><strong>NOTE 1:</strong> The port number (i.e. 2110) must match with the output of OpenCPU in your RStudio session.
This can differ everytime you start the application.</p>

<p><strong>NOTE 2:</strong> The application stores the most important data resulting from your analysis (i.e. clustering results and tags).
An analysis can therefore be splitted into several sessions.
The initial preparations are not required after the first (assuming binning of the same data set is continued).
To continue a previously stopped session, just start R(Studio), load the OpenCPU library as shown above, and point your browser to the correct url.</p>

<h2 id="re-assembly-and-verification">Re-assembly and verification</h2>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Interactive Contig Binning</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li>Interactive Contig Binning</li>
          <li><a href="mailto:bertjan.broeksema@list.lu">bertjan.broeksema@list.lu</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/bbroeksema">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">bbroeksema</span>
            </a>
          </li>
          

          
          <li>
            <a href="https://twitter.com/looslycoupled">
              <span class="icon  icon--twitter">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M15.969,3.058c-0.586,0.26-1.217,0.436-1.878,0.515c0.675-0.405,1.194-1.045,1.438-1.809
                  c-0.632,0.375-1.332,0.647-2.076,0.793c-0.596-0.636-1.446-1.033-2.387-1.033c-1.806,0-3.27,1.464-3.27,3.27 c0,0.256,0.029,0.506,0.085,0.745C5.163,5.404,2.753,4.102,1.14,2.124C0.859,2.607,0.698,3.168,0.698,3.767 c0,1.134,0.577,2.135,1.455,2.722C1.616,6.472,1.112,6.325,0.671,6.08c0,0.014,0,0.027,0,0.041c0,1.584,1.127,2.906,2.623,3.206 C3.02,9.402,2.731,9.442,2.433,9.442c-0.211,0-0.416-0.021-0.615-0.059c0.416,1.299,1.624,2.245,3.055,2.271 c-1.119,0.877-2.529,1.4-4.061,1.4c-0.264,0-0.524-0.015-0.78-0.046c1.447,0.928,3.166,1.469,5.013,1.469 c6.015,0,9.304-4.983,9.304-9.304c0-0.142-0.003-0.283-0.009-0.423C14.976,4.29,15.531,3.714,15.969,3.058z"/>
                </svg>
              </span>

              <span class="username">looslycoupled</span>
            </a>
          </li>
          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text">This site provides data, scripts and a tutorial to replicate the contig binning results presented in [](PAPER titel).
</p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
